### PR TITLE
bpfman-operator: Add Config resource and update reconciler and tests 

### PR DESCRIPTION
**Pull request by andreaskaris**
## Summary

This PR introduces a new Config CRD to replace ConfigMap usage for managing cluster-scoped resources, addressing ownership reference limitations where namespace-scoped ConfigMaps cannot own cluster-scoped resources like CSIDriver. The change includes lifecycle testing and reconciliation improvements.

## Changes

- **Add Config CRD**: New custom resource definition with proper API types and validation
- **Replace ConfigMap controller**: Migrate from configmap.go to config.go with enhanced reconciliation logic
- **Fix ownership references**: Enable proper cascading deletion for cluster-scoped resources
- **Add finalizer protection**: Prevent race conditions during Config deletion with proper finalizer management
- **Improve reconciliation**: Use semantic comparison to prevent unnecessary resource updates and reconciliation loops
- **Enhanced error handling**: Add structured logging and better error reporting in reconciler
- **Update deployment config**: Modify bpfman-deployment configuration to use new Config resource
- **Add lifecycle testing**: Comprehensive integration test suite covering resource creation, updates, and restoration scenarios
- **Refactor OpenShift detection**: Extract OpenShift detection logic to internal package for better testability
- **Generate client code**: Auto-generated deepcopy, client, and informer code for Config resource
- **Sync deployment manifests**: Update DaemonSet and metrics-proxy YAML with missing Kubernetes default fields

## Technical Details

The transition from ConfigMap to Config CRD resolves the fundamental issue where namespace-scoped resources cannot establish  owner references with cluster-scoped resources. The new Config controller provides the same functionality with proper resource  lifecycle management through Kubernetes garbage collection.

Key improvements include:
- **Finalizer-based deletion**: Ensures coordinated cleanup preventing race conditions between operator and garbage collection
- **Semantic comparison**: Prevents reconciliation loops by only updating resources when meaningful changes occur
- **Field validation**: API server validation for configuration values through CRD schema
- **Lifecycle testing**: End-to-end test coverage for Config resource operations

The finalizer ensures a safe deletion sequence: user deletes Config → deletion timestamp set → operator stops creating/updating resources → finalizer removed → garbage collection proceeds via owner references.

## Testing

All existing functionality is preserved with enhanced test coverage for edge cases and error scenarios.

Tested with: (/)

* ci integration test:
```
export BPFMAN_AGENT_IMG=quay.io/bpfman/bpfman-agent:int-test; export BPFMAN_OPERATOR_IMG=quay.io/bpfman/bpfman-operator:int-test; make build-images; make test-integration; unset BPFMAN_AGENT_IMG; unset BPFMAN_OPERATOR_IMG
```

* local integration test: (/)
```
make run-on-kind && make test-integration-local
```

* unit tests: (/)
```
make test
```

* linter (/)
```
make lint
```

* bundle (/)
```
make bundle
```

* openshift deployment: (/)
```
export OCI_BIN=podman
export BPFMAN_AGENT_IMG=quay.io/akaris/bpfman-agent:reconcile
export BPFMAN_OPERATOR_IMG=quay.io/akaris/bpfman-operator:reconcile
make build-images 
make push-images 
make deploy-openshift
make undeploy-openshift
```

* lifecycle tests:

- [x] kind
```
make run-on-kind
make test-lifecycle-local
```

- [x] openshift
```
export BPFMAN_AGENT_IMG=quay.io/akaris/bpfman-agent:reconcile
export BPFMAN_OPERATOR_IMG=quay.io/akaris/bpfman-operator:reconcile
make deploy-openshift
make test-lifecycle-local
```

```
[akaris@workstation bpfman-operator (reconcile)]$ make test-lifecycle-local
GOFLAGS="-tags=integration_tests" go test -count=1 -race -v ./test/lifecycle
=== RUN   TestLifecycle
INFO: Determine if this is OpenShift, if so get the OpenShift client
{"level":"info","ts":"2025-08-22T04:39:23+02:00","msg":"detected platform version","PlatformVersion":"v1.31.7"}
{"level":"info","ts":"2025-08-22T04:39:23+02:00","msg":"route.openshift.io found in apis, platform is OpenShift"}
    lifecycle_test.go:94: Running: TestGetCurrentConfig
    lifecycle_test.go:115: Running: TestEnsureResources
    lifecycle_test.go:121: Running: TestResourceDeletion
    lifecycle_test.go:127: Running: TestResourceModification
    lifecycle_test.go:133: Running: TestConfigCascadingDeletion
    lifecycle_test.go:139: Running: TestConfigRecreationWithModifiedSettings
    lifecycle_test.go:102: Restoring original config
--- PASS: TestLifecycle (46.19s)
PASS
ok  	github.com/bpfman/bpfman-operator/test/lifecycle	47.235s
[akaris@workstation bpfman-operator (reconcile)]$ 
```

## Further information

* https://github.com/kubernetes/community/blob/master/contributors/devel/sig-api-machinery/controllers.md#guidelines
> Consider using owner references for resources that result in the creation of other resources (eg. a ReplicaSet results in creating Pods). Thus you ensure that children resources are going to be garbage-collected once a resource managed by your controller is deleted. For more information on owner references, read more [here](https://git.k8s.io/design-proposals-archive/api-machinery/controller-ref.md).

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -466,7 +466,7 @@ kind-reload-images: load-images-kind ## Reload locally build images into a kind
 	kubectl rollout restart deployment bpfman-operator -n bpfman
 
 .PHONY: run-on-kind
-run-on-kind: kustomize setup-kind build-images load-images-kind deploy ## Kind Deploy runs the bpfman-operator on a local kind cluster using local builds of bpfman, bpfman-agent, and bpfman-operator
+run-on-kind: kustomize setup-kind build-images load-images-kind install deploy ## Kind Deploy runs the bpfman-operator on a local kind cluster using local builds of bpfman, bpfman-agent, and bpfman-operator
```
Comment (andreaskaris) [inline Makefile:469]: install is needed because CRD must be created before deploy of Config

File: controllers/bpfman-operator/config.go [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+		}
+		r.Logger.Error(err, "Failed to get object",
+			"type", resource.GetObjectKind(), "namespace", resource.GetNamespace(), "name", resource.GetName())
+		return err
+	}
+
+	r.Logger.Info("Reconciling bpfman object",
+		"type", resource.GetObjectKind(), "namespace", resource.GetNamespace(), "name", resource.GetName())
+	// Let the Kubernetes API server determine what needs updating.
+	// DeepEqual comparison is unreliable due to server-side default field mutations.
```
Comment (andreaskaris) [inline controllers/bpfman-operator/config.go:364]: To be honest, I mainly dropped the DeepEqual because the .Spec field doesn't play well with generics (there's no GetSpec() given that all specs are of different types). Be it as it may, I don't believe that updating resources is that costly given that it's just a few resources and fingers crossed they're stable .. Should this yield unwanted changes to the DS and pod redeployments at each reconcile of unrelated resources -> something more targeted than a DeepEqual would likely be in order

Comment (andreaskaris) [reply controllers/bpfman-operator/config.go:364]: Something like:
```
[akaris@workstation bpfman-operator (reconcile)]$ git diff
diff --git a/controllers/bpfman-operator/config.go b/controllers/bpfman-operator/config.go
index 7a5f5f61..d90e691b 100644
--- a/controllers/bpfman-operator/config.go
+++ b/controllers/bpfman-operator/config.go
@@ -134,7 +134,7 @@ func (r *BpfmanConfigReconciler) reconcileCM(ctx context.Context, bpfmanConfig *
                        internal.BpfmanLogLevel:      bpfmanConfig.Spec.LogLevel,
                },
        }
-       return assureResource(ctx, r, bpfmanConfig, cm)
+       return assureResource(ctx, r, bpfmanConfig, cm, func(existing, desired *corev1.ConfigMap) bool { return true })
 }
 
 func (r *BpfmanConfigReconciler) reconcileCSIDriver(ctx context.Context, bpfmanConfig *v1alpha1.Config) error {
@@ -144,7 +144,7 @@ func (r *BpfmanConfigReconciler) reconcileCSIDriver(ctx context.Context, bpfmanC
        if err != nil {
                return err
        }
-       return assureResource(ctx, r, bpfmanConfig, csiDriver)
+       return assureResource(ctx, r, bpfmanConfig, csiDriver, func(existing, desired *storagev1.CSIDriver) bool { return true })
 }
 
 func (r *BpfmanConfigReconciler) reconcileSCC(ctx context.Context, bpfmanConfig *v1alpha1.Config) error {
@@ -159,7 +159,7 @@ func (r *BpfmanConfigReconciler) reconcileSCC(ctx context.Context, bpfmanConfig
                if err != nil {
                        return err
                }
-               return assureResource(ctx, r, bpfmanConfig, bpfmanRestrictedSCC)
+               return assureResource(ctx, r, bpfmanConfig, bpfmanRestrictedSCC, func(existing, desired *osv1.SecurityContextConstraints) bool { return true })
        }
        return nil
 }
@@ -172,7 +172,7 @@ func (r *BpfmanConfigReconciler) reconcileStandardDS(ctx context.Context, bpfman
                return err
        }
        configureBpfmanDs(bpfmanDS, bpfmanConfig)
-       return assureResource(ctx, r, bpfmanConfig, bpfmanDS)
+       return assureResource(ctx, r, bpfmanConfig, bpfmanDS, func(existing, desired *appsv1.DaemonSet) bool { return true })
 }
 
 func (r *BpfmanConfigReconciler) reconcileMetricsProxyDS(ctx context.Context, bpfmanConfig *v1alpha1.Config) error {
@@ -184,7 +184,7 @@ func (r *BpfmanConfigReconciler) reconcileMetricsProxyDS(ctx context.Context, bp
                return err
        }
        configureMetricsProxyDs(metricsProxyDS, bpfmanConfig, r.IsOpenshift)
-       return assureResource(ctx, r, bpfmanConfig, metricsProxyDS)
+       return assureResource(ctx, r, bpfmanConfig, metricsProxyDS, func(existing, desired *appsv1.DaemonSet) bool { return true })
 }
 
 // resourcePredicate creates a predicate that filters events based on resource name.
@@ -333,7 +333,7 @@ func load[T client.Object](t T, path, name string) (T, error) {
 // assureResource ensures a Kubernetes resource exists and is up to date.
 // Creates the resource if it doesn't exist, otherwise updates it to match the desired state.
 func assureResource[T client.Object](ctx context.Context, r *BpfmanConfigReconciler,
-       bpfmanConfig *v1alpha1.Config, resource T) error {
+       bpfmanConfig *v1alpha1.Config, resource T, needsUpdate func(existing T, desired T) bool) error {
        if err := ctrl.SetControllerReference(bpfmanConfig, resource, r.Scheme); err != nil {
                return err
        }
@@ -358,15 +358,15 @@ func assureResource[T client.Object](ctx context.Context, r *BpfmanConfigReconci
                return err
        }
 
-       r.Logger.Info("Reconciling bpfman object",
-               "type", resource.GetObjectKind(), "namespace", resource.GetNamespace(), "name", resource.GetName())
-       // Let the Kubernetes API server determine what needs updating.
-       // DeepEqual comparison is unreliable due to server-side default field mutations.
-       resource.SetResourceVersion(existingResource.GetResourceVersion())
-       if err := r.Update(ctx, resource); err != nil {
-               r.Logger.Error(err, "Failed reconciling object",
+       if needsUpdate(existingResource, resource) {
+               r.Logger.Info("Updating object",
                        "type", resource.GetObjectKind(), "namespace", resource.GetNamespace(), "name", resource.GetName())
-               return err
+               resource.SetResourceVersion(existingResource.GetResourceVersion())
+               if err := r.Update(ctx, resource); err != nil {
+                       r.Logger.Error(err, "Failed updating object",
+                               "type", resource.GetObjectKind(), "namespace", resource.GetNamespace(), "name", resource.GetName())
+                       return err
+               }
        }
 
        return nil
```

File: controllers/bpfman-operator/config.go [hunk 3]
```diff
@@ -101,6 +103,25 @@ func (r *BpfmanConfigReconciler) Reconcile(ctx context.Context, req ctrl.Request
 		return ctrl.Result{}, err
 	}
 
+	// Check if Config is being deleted first to prevent race
+	// conditions.
+	if !bpfmanConfig.DeletionTimestamp.IsZero() {
+		return r.handleDeletion(ctx, bpfmanConfig)
```
[the remark this answers is not in the document]
Comment (andreaskaris) [reply controllers/bpfman-operator/config.go:109]: I pushed this minor change here for the logging: https://github.com/bpfman/bpfman-operator/compare/fda909b0ad3aef26c1e3fdd90e7a63ee2b3b0d7a..d4320a3935a035472a88a8c82744d7da57054224

Other than that, if you're o.k. with it, let's merge it

**Comment by frobware** [comment]
@andreaskaris - we need https://github.com/frobware/bpfman-operator/commit/4b9e667a69b77926c97c02e8d82a9f28666c57a7, otherwise the image value substitutions no longer work. Admittedly, I've only tested the `deploy` and `deploy-openshift` targets.

**Comment by frobware** [comment]
> @andreaskaris - we need [frobware@4b9e667](https://github.com/frobware/bpfman-operator/commit/4b9e667a69b77926c97c02e8d82a9f28666c57a7), otherwise the image value substitutions no longer work. Admittedly, I've only tested the `deploy` and `deploy-openshift` targets.

Actually, my change is only mostly write. I keep forgetting that kustomize (given our setup) will prefix with `bpfman-`. So the name for the config object is bpfman-config. Let me adjust my patch.

**Comment by andreaskaris** [comment]
Ok, I added it to the PR - thanks for catching that!

**Comment by frobware** [comment]
I also think we would benefit from https://github.com/frobware/bpfman-operator/commit/28afe1834dc2f6f879fee771566cf4d4ea67abaa. Makes undeploy much simpler.

**Comment by andreaskaris** [comment]
Great, I added that, as well. I also had to throw in https://github.com/bpfman/bpfman-operator/pull/461/commits/c91717705ea164a09a703f4973b94a395c551edf for github CI to get kustomize, otherwise the integration tests failed because your patches add a dependency on KUSTOMIZE for the test-integration target

**Comment by frobware** [comment]
I ran:

```sh
% make test-lifecycle-local
GOFLAGS="-tags=integration_tests" go test -count=1 -race -v ./test/lifecycle
=== RUN   TestLifecycle
INFO: Determine if this is OpenShift
{"level":"info","ts":"2025-08-26T15:15:16+01:00","msg":"detected platform version","PlatformVersion":"v1.32.6"}
{"level":"info","ts":"2025-08-26T15:15:16+01:00","msg":"route.openshift.io found in apis, platform is OpenShift"}
    lifecycle_test.go:105: Running: TestGetCurrentConfig
    lifecycle_test.go:124: Running: TestEnsureResources
    lifecycle_test.go:130: Running: TestResourceDeletion
    lifecycle_test.go:136: Running: TestResourceModification
    lifecycle_test.go:253: Checking that config map was correctly reconciled
    lifecycle_test.go:298: Checking that CSI was correctly reconciled
    lifecycle_test.go:351: Checking that standard ds was correctly reconciled
    lifecycle_test.go:397: Checking that metrics ds was correctly reconciled
    lifecycle_test.go:436: Checking that standard metrics ds was correctly reconciled
    lifecycle_test.go:142: Running: TestConfigCascadingDeletion
    lifecycle_test.go:148: Running: TestConfigRecreationWithModifiedSettings
    lifecycle_test.go:114: Restoring original config
--- PASS: TestLifecycle (58.50s)
PASS
ok      github.com/bpfman/bpfman-operator/test/lifecycle        59.539s
```

should we extend this to mutate the configmap and verify that its contents get overwritten with whatever is set in the config?

**Comment by andreaskaris** [comment]
> I ran:
> 
> ```shell
> % make test-lifecycle-local
> GOFLAGS="-tags=integration_tests" go test -count=1 -race -v ./test/lifecycle
> === RUN   TestLifecycle
> INFO: Determine if this is OpenShift
> {"level":"info","ts":"2025-08-26T15:15:16+01:00","msg":"detected platform version","PlatformVersion":"v1.32.6"}
> {"level":"info","ts":"2025-08-26T15:15:16+01:00","msg":"route.openshift.io found in apis, platform is OpenShift"}
>     lifecycle_test.go:105: Running: TestGetCurrentConfig
>     lifecycle_test.go:124: Running: TestEnsureResources
>     lifecycle_test.go:130: Running: TestResourceDeletion
>     lifecycle_test.go:136: Running: TestResourceModification
>     lifecycle_test.go:253: Checking that config map was correctly reconciled
>     lifecycle_test.go:298: Checking that CSI was correctly reconciled
>     lifecycle_test.go:351: Checking that standard ds was correctly reconciled
>     lifecycle_test.go:397: Checking that metrics ds was correctly reconciled
>     lifecycle_test.go:436: Checking that standard metrics ds was correctly reconciled
>     lifecycle_test.go:142: Running: TestConfigCascadingDeletion
>     lifecycle_test.go:148: Running: TestConfigRecreationWithModifiedSettings
>     lifecycle_test.go:114: Restoring original config
> --- PASS: TestLifecycle (58.50s)
> PASS
> ok      github.com/bpfman/bpfman-operator/test/lifecycle        59.539s
> ```
> 
> should we extend this to mutate the configmap and verify that its contents get overwritten with whatever is set in the config?

I'm already testing this here, if I understand you correctly:
* https://github.com/bpfman/bpfman-operator/pull/461/commits/a8d5c89b1cd4786f7a733b2d280f68b9eb83ce01#diff-0be51b266e4492426d513bee31a3f0416d802d0ff8032136bc9f6b1b31d76ff0R136
* https://github.com/bpfman/bpfman-operator/pull/461/commits/a8d5c89b1cd4786f7a733b2d280f68b9eb83ce01#diff-0be51b266e4492426d513bee31a3f0416d802d0ff8032136bc9f6b1b31d76ff0R225

**Comment by frobware** [comment]
> > I ran:
> > ```shell
> > % make test-lifecycle-local
> > GOFLAGS="-tags=integration_tests" go test -count=1 -race -v ./test/lifecycle
> > === RUN   TestLifecycle
> > INFO: Determine if this is OpenShift
> > {"level":"info","ts":"2025-08-26T15:15:16+01:00","msg":"detected platform version","PlatformVersion":"v1.32.6"}
> > {"level":"info","ts":"2025-08-26T15:15:16+01:00","msg":"route.openshift.io found in apis, platform is OpenShift"}
> >     lifecycle_test.go:105: Running: TestGetCurrentConfig
> >     lifecycle_test.go:124: Running: TestEnsureResources
> >     lifecycle_test.go:130: Running: TestResourceDeletion
> >     lifecycle_test.go:136: Running: TestResourceModification
> >     lifecycle_test.go:253: Checking that config map was correctly reconciled
> >     lifecycle_test.go:298: Checking that CSI was correctly reconciled
> >     lifecycle_test.go:351: Checking that standard ds was correctly reconciled
> >     lifecycle_test.go:397: Checking that metrics ds was correctly reconciled
> >     lifecycle_test.go:436: Checking that standard metrics ds was correctly reconciled
> >     lifecycle_test.go:142: Running: TestConfigCascadingDeletion
> >     lifecycle_test.go:148: Running: TestConfigRecreationWithModifiedSettings
> >     lifecycle_test.go:114: Restoring original config
> > --- PASS: TestLifecycle (58.50s)
> > PASS
> > ok      github.com/bpfman/bpfman-operator/test/lifecycle        59.539s
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > should we extend this to mutate the configmap and verify that its contents get overwritten with whatever is set in the config?
> 
> I'm already testing this here, if I understand you correctly:
> 
> * [a8d5c89#diff-0be51b266e4492426d513bee31a3f0416d802d0ff8032136bc9f6b1b31d76ff0R136](https://github.com/bpfman/bpfman-operator/commit/a8d5c89b1cd4786f7a733b2d280f68b9eb83ce01#diff-0be51b266e4492426d513bee31a3f0416d802d0ff8032136bc9f6b1b31d76ff0R136)
> * [a8d5c89#diff-0be51b266e4492426d513bee31a3f0416d802d0ff8032136bc9f6b1b31d76ff0R225](https://github.com/bpfman/bpfman-operator/commit/a8d5c89b1cd4786f7a733b2d280f68b9eb83ce01#diff-0be51b266e4492426d513bee31a3f0416d802d0ff8032136bc9f6b1b31d76ff0R225)
> 
> Or do you mean here, by inverting TestConfigRecreationWithModifiedSettings with TestConfigCascadingDeletion?
> 
> * [a8d5c89#diff-0be51b266e4492426d513bee31a3f0416d802d0ff8032136bc9f6b1b31d76ff0R148](https://github.com/bpfman/bpfman-operator/commit/a8d5c89b1cd4786f7a733b2d280f68b9eb83ce01#diff-0be51b266e4492426d513bee31a3f0416d802d0ff8032136bc9f6b1b31d76ff0R148)

Sorry - let me be clearer.

If you inadvertently mutate the config map (i.e., the old way), do we have a test in place to verify whether any modifications are overwritten with the values from the top-level Config configuration? In other words, do we have a mechanism to prevent changes to the config map?

**Comment by frobware** [comment]
> > If you inadvertently mutate the config map (i.e., the old way), do we have a test in place to verify whether any modifications are overwritten with the values from the top-level Config configuration? In other words, do we have a mechanism to prevent changes to the config map?
> 
> If you make a change to the configmap, the Config controller will revert the change. That's tested here:
> 
> * https://github.com/bpfman/bpfman-operator/blob/c91717705ea164a09a703f4973b94a395c551edf/test/lifecycle/lifecycle_test.go#L225

Great, this what is was looking for.

> 
> > do we have a mechanism to prevent changes to the config map?
> 
> To prevent changes per se, no. I don't think that works in kubernetes, anyway - unless you deploy a webhook in some shape or form? If you make a change, the change will be accepted, but the reconciler will immediately revert it, and that's tested in the code above -^

yes, fine and expected.

File: controllers/bpfman-operator/config.go [hunk 4]
```diff
@@ -0,0 +1,441 @@
...
+}
+
+func healthProbeAddress(healthProbePort int) string {
+	if healthProbePort <= 0 || healthProbePort > 65535 {
+		return ""
+	}
+	return fmt.Sprintf(":%d", healthProbePort)
+}
+
+func pointer[T any](t T) *T {
```
Comment (frobware) [inline controllers/bpfman-operator/config.go:439]: All usages of `pointer` can be `ptr.To`.

```diff
diff --git a/controllers/bpfman-operator/config.go b/controllers/bpfman-operator/config.go
index 9ce8bca5..d45709d6 100644
--- a/controllers/bpfman-operator/config.go
+++ b/controllers/bpfman-operator/config.go
@@ -333,7 +333,7 @@ func configureMetricsProxyDs(staticMetricsProxyDS *appsv1.DaemonSet, config *v1a
 				VolumeSource: corev1.VolumeSource{
 					Secret: &corev1.SecretVolumeSource{
 						SecretName:  "agent-metrics-tls",
-						DefaultMode: pointer(int32(420)),
+						DefaultMode: ptr.To(int32(420)),
 					},
 				},
 			},
diff --git a/controllers/bpfman-operator/config_test.go b/controllers/bpfman-operator/config_test.go
index 5aa78370..35005e75 100644
--- a/controllers/bpfman-operator/config_test.go
+++ b/controllers/bpfman-operator/config_test.go
@@ -33,6 +33,7 @@ import (
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/types"
 	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/utils/ptr"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/client/fake"
 	logf "sigs.k8s.io/controller-runtime/pkg/log"
@@ -338,8 +339,8 @@ func testAllObjectsPresent(ctx context.Context, cl client.Client, bpfmanConfig *
 	}
 	requiredFields := map[string]*string{
 		internal.BpfmanTOML:          nil,
-		internal.BpfmanAgentLogLevel: pointer(logLevel),
-		internal.BpfmanLogLevel:      pointer(logLevel),
+		internal.BpfmanAgentLogLevel: ptr.To(logLevel),
+		internal.BpfmanLogLevel:      ptr.To(logLevel),
 	}
 	if err := verifyCM(bpfmanCM, requiredFields); err != nil {
 		return err
@@ -500,9 +501,9 @@ func modifyObjects(ctx context.Context, cl client.Client, isOpenShift bool) erro
 		return err
 	}
 	if csiDriver.Spec.AttachRequired == nil || !*csiDriver.Spec.AttachRequired {
-		csiDriver.Spec.AttachRequired = pointer(true)
+		csiDriver.Spec.AttachRequired = ptr.To(true)
 	} else {
-		csiDriver.Spec.AttachRequired = pointer(false)
+		csiDriver.Spec.AttachRequired = ptr.To(false)
 	}
 	if err := cl.Update(ctx, csiDriver); err != nil {
 		return err
```

File: controllers/bpfman-operator/config.go [hunk 5]
```diff
@@ -0,0 +1,441 @@
...
+					},
+				},
+			},
+		)
+	}
+}
+
+// load reads a Kubernetes resource manifest from the specified file path and deserializes it.
+// Sets the resource name and returns the loaded object.
+func load[T client.Object](t T, path, name string) (T, error) {
```
Comment (frobware) [inline controllers/bpfman-operator/config.go:346]: wraps all errors paths with context showing which resource and file path failed to load. This will make debugging much easier when manifest loading fails... (if ever). But if it does, I'd like to see _which_ file.

```go
func load[T client.Object](t T, path, name string) (T, error) {
	file, err := os.Open(path)
	if err != nil {
		return t, fmt.Errorf("load %s from %s: %w", name, path, err)
	}
	defer file.Close()

	b, err := io.ReadAll(file)
	if err != nil {
		return t, fmt.Errorf("load %s from %s: %w", name, path, err)
	}

	decode := scheme.Codecs.UniversalDeserializer().Decode
	obj, _, err := decode(b, nil, t)
	if err != nil {
		return t, fmt.Errorf("load %s from %s: %w", name, path, err)
	}

	t = obj.(T)
	t.SetName(name)

	return t, nil
}
```

Comment (andreaskaris) [reply controllers/bpfman-operator/config.go:346]: thanks for the suggestion, done

File: apis/v1alpha1/config_types.go [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+// AgentSpec defines the desired state of the bpfman agent.
+type AgentSpec struct {
+	// HealthProbePort holds the health probe bind port for the bpfman agent.
+	// +optional
+	// +kubebuilder:default=8175
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:validation:Maximum=65535
+	HealthProbePort int `json:"healthProbeAddr"`
```
Comment (dave-tucker) [inline apis/v1alpha1/config_types.go:69]: nit: the JSON fieldName is different from the struct Name.

**Review by dave-tucker — APPROVED** [review]
One small nit, but otherwise LGTM @andreaskaris.
As a follow up, we'll probably need to write some documentation - which I believe is still in bpfman/bpfman - since it's a breaking change. It is however a lot cleaner, kudos!
